Added tensor size warning to F.mse_loss()

mc-robinson opened the pull request:
To address the issue of broadcasting giving the wrong result in `nn.MSELoss()` as mentioned here https://github.com/pytorch/pytorch/issues/16045 . In particular, the issue often arises when computing the loss between tensors with shapes (n, 1) and (n,)

ailzhang commented:
@mc-robinson  Can you add a test to assert this warning shows up? Thanks!

mc-robinson commented:
Sure. But sorry, I'm having trouble seeing where exactly the test should go? `test/test_nn.py` and `test/common_nn.py` seem like the best candidates I can see

ailzhang commented:
Somewhere here should be good. https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L4332
Btw if we are disabling it for `mse_loss` it might be good to disable for all other applicable cases as well. I don't know much detail if broadcasting is useful in any of other loss functions. If not, we should probably just warn all functions that currently do broacast. 
cc: @soumith  do you know in any context broadcasting is helpful?

SsnL commented:
In a lot of cases broadcasting is helpful though, e.g., multiple prediction of the same target. I think what you want to do is to warn only if the input is going to be expanded.

mc-robinson commented:
I get that broadcasting is useful for loss functions like `F.cross_entropy`, but have you found it useful for `nn.MSELoss()` @SsnL ? 

And only warning if the input is going to be expanded misses the main issue in #16045 . In that case, the input needs to be squeezed or target must be unsqueezed.

mc-robinson commented:
Looks like there might need to be a more extensive look of how each loss function deals with broadcasting to build in some consistency -- I might look into it tomorrow. FWIW, fastai seems to have partially dealt with this issue by creating a "FlattenedLoss()" class that is used as a default https://github.com/fastai/fastai/blob/master/fastai/layers.py#L247

mc-robinson committed:
added test

mc-robinson committed:
Merge pull request #2 from pytorch/master

update fork

mc-robinson commented:
I added the necessary test, and also added `stacklevel=2` to a couple of warning messages so that they now actually detail the last function call that triggered the warning. Before, the message was just a repeat of the last line of the warning and was essentially useless.

This certainly doesn't fix everything, but at least gets a warning to people for now.

Also, I stupidly forgot to update my fork before committing at first, so please let me know if that caused any problems. Still trying to get the hang of this!

apaszke commented on torch/nn/functional.py:
```diff
@@ -2204,6 +2205,11 @@ def mse_loss(input, target, size_average=None, reduce=None, reduction='mean'):
 
     See :class:`~torch.nn.MSELoss` for details.
     """
+    if not (target.size() == input.size()):
```
It would be more readable and efficient to do `target.size() != input.size()`. Can we please use that here and above as well?

soumith committed:
not -> !=

soumith committed:
fix lint

soumith committed:
fix lint

soumith reviewed (approved):
i made the minor requested changes. thanks a lot @mc-robinson
